### PR TITLE
[Feature] Implement application deletion

### PR DESCRIPTION
**Pull request by leogjhuang**
## Notion ticket link

[Implement application deletion](https://www.notion.so/19-Implement-applications-Deletion-28701aa8b7964893a10e366d2183f2cf)

## Implementation description
* Application/request deletion feature
* Designed to maintain consistency with #310

## Notes
* Database schema changes will need to be enabled to allow for on delete cascade

### Before deleting Application 3
```
rcd=# select * from replacement_applications where application_id = 3;
 application_id | reason |     lost_timestamp     | lost_location | stolen_police_file_number | stolen_jurisdiction | stolen_police_officer_name |                      event_description                      |         created_at         |         updated_at         
----------------+--------+------------------------+---------------+---------------------------+---------------------+----------------------------+-------------------------------------------------------------+----------------------------+----------------------------
              3 | LOST   | 2020-01-03 00:00:00+00 | DC Library    |                           |                     |                            | I left my APP on the table and when I came back it was gone | 2023-09-03 09:26:59.326+00 | 2023-09-03 09:26:59.326+00
(1 row)

rcd=# select * from permits where application_id = 3;
 rcd_permit_id |   type    | expiry_date | active | application_id | applicant_id |         created_at         |         updated_at         
---------------+-----------+-------------+--------+----------------+--------------+----------------------------+----------------------------
             2 | PERMANENT | 2025-01-01  | t      |              3 |            3 | 2023-09-03 09:26:59.377+00 | 2023-09-03 09:26:59.377+00
(1 row)

rcd=# select * from application_processing where id = 3;
 id | status  | rejected_reason | app_number | app_number_employee_id |   app_number_updated_at    | app_holepunched | app_holepunched_employee_id | app_holepunched_updated_at | wallet_card_created | wallet_card_created_employee_id | wallet_card_created_updated_at | review_request_completed | review_request_completed_employee_id | review_request_completed_updated_at | invoice_number | documents_s3_object_key | documents_url_employee_id |  documents_url_updated_at  | app_mailed | app_mailed_employee_id |   app_mailed_updated_at    | payment_refunded | payment_refunded_employee_id | payment_refunded_updated_at |         created_at         |         updated_at         
----+---------+-----------------+------------+------------------------+----------------------------+-----------------+-----------------------------+----------------------------+---------------------+---------------------------------+--------------------------------+--------------------------+--------------------------------------+-------------------------------------+----------------+-------------------------+---------------------------+----------------------------+------------+------------------------+----------------------------+------------------+------------------------------+-----------------------------+----------------------------+----------------------------
  3 | PENDING |                 |            |                        | 2023-09-03 09:26:59.326+00 | f               |                             | 2023-09-03 09:26:59.326+00 | f                   |                                 | 2023-09-03 09:26:59.326+00     | f                        |                                      | 2023-09-03 09:26:59.326+00          |                |                         |                           | 2023-09-03 09:26:59.326+00 | f          |                        | 2023-09-03 09:26:59.326+00 | f                |                              | 2023-09-03 09:26:59.326+00  | 2023-09-03 09:26:59.326+00 | 2023-09-03 09:26:59.326+00
(1 row)

rcd=# select * from applications where id = 3;
 id | first_name | middle_name | last_name |   phone    |          email           | receive_email_updates |  address_line_1   | address_line_2 |   city    | province | country | postal_code | permit_type | payment_method | processing_fee | donation_amount | paid_through_shopify | shopify_payment_status | shopify_confirmation_number | shopify_order_number | shipping_address_same_as_home_address | shipping_full_name | shipping_address_line_1 | shipping_address_line_2 | shipping_city | shipping_province | shipping_country | shipping_postal_code | billing_address_same_as_home_address | billing_full_name | billing_address_line_1 | billing_address_line_2 | billing_city | billing_province | billing_country | billing_postal_code |    type     | notes | applicant_id | application_processing_id |         created_at         |         updated_at         
----+------------+-------------+-----------+------------+--------------------------+-----------------------+-------------------+----------------+-----------+----------+---------+-------------+-------------+----------------+----------------+-----------------+----------------------+------------------------+-----------------------------+----------------------+---------------------------------------+--------------------+-------------------------+-------------------------+---------------+-------------------+------------------+----------------------+--------------------------------------+-------------------+------------------------+------------------------+--------------+------------------+-----------------+---------------------+-------------+-------+--------------+---------------------------+----------------------------+----------------------------
  3 | Applicant  |             | Three     | 0987654321 | applicantthree@gmail.com | f                     | 789 Vancouver Rd. |                | Vancouver | BC       | Canada  | B1C2D3      | PERMANENT   | CHEQUE         |         $31.00 |         $100.00 | f                    |                        |                             |                      | t                                     |                    |                         |                         |               |                   |                  |                      | t                                    |                   |                        |                        |              |                  |                 |                     | REPLACEMENT |       |            3 |                         3 | 2023-09-03 09:26:59.326+00 | 2023-09-03 09:26:59.326+00
(1 row)
```
### Demo
https://github.com/uwblueprint/richmond-centre-for-disability/assets/77544794/a24f9f3a-1f2d-4e5a-98ff-a133845da141
### After deleting Application 3
```
rcd=# select * from replacement_applications where application_id = 3;
tion_id = 3;
select * from application_processing  application_id | reason | lost_timestamp | lost_location | stolen_police_file_number | stolen_jurisdiction | stolen_police_officer_name | event_description | created_at | updated_at 
----------------+--------+----------------+---------------+---------------------------+---------------------+----------------------------+-------------------+------------+------------
(0 rows)

rcd=# select * from permits where application_id = 3;
 rcd_permit_id | type | expiry_date | active | application_id | applicant_id | created_at | updated_at 
---------------+------+-------------+--------+----------------+--------------+------------+------------
(0 rows)

rcd=# select * from application_processing where id = 3;
 id | status | rejected_reason | app_number | app_number_employee_id | app_number_updated_at | app_holepunched | app_holepunched_employee_id | app_holepunched_updated_at | wallet_card_created | wallet_card_created_employee_id | wallet_card_created_updated_at | review_request_completed | review_request_completed_employee_id | review_request_completed_updated_at | invoice_number | documents_s3_object_key | documents_url_employee_id | documents_url_updated_at | app_mailed | app_mailed_employee_id | app_mailed_updated_at | payment_refunded | payment_refunded_employee_id | payment_refunded_updated_at | created_at | updated_at 
----+--------+-----------------+------------+------------------------+-----------------------+-----------------+-----------------------------+----------------------------+---------------------+---------------------------------+--------------------------------+--------------------------+--------------------------------------+-------------------------------------+----------------+-------------------------+---------------------------+--------------------------+------------+------------------------+-----------------------+------------------+------------------------------+-----------------------------+------------+------------
(0 rows)

rcd=# select * from applications where id = 3;
 id | first_name | middle_name | last_name | phone | email | receive_email_updates | address_line_1 | address_line_2 | city | province | country | postal_code | permit_type | payment_method | processing_fee | donation_amount | paid_through_shopify | shopify_payment_status | shopify_confirmation_number | shopify_order_number | shipping_address_same_as_home_address | shipping_full_name | shipping_address_line_1 | shipping_address_line_2 | shipping_city | shipping_province | shipping_country | shipping_postal_code | billing_address_same_as_home_address | billing_full_name | billing_address_line_1 | billing_address_line_2 | billing_city | billing_province | billing_country | billing_postal_code | type | notes | applicant_id | application_processing_id | created_at | updated_at 
----+------------+-------------+-----------+-------+-------+-----------------------+----------------+----------------+------+----------+---------+-------------+-------------+----------------+----------------+-----------------+----------------------+------------------------+-----------------------------+----------------------+---------------------------------------+--------------------+-------------------------+-------------------------+---------------+-------------------+------------------+----------------------+--------------------------------------+-------------------+------------------------+------------------------+--------------+------------------+-----------------+---------------------+------+-------+--------------+---------------------------+------------+------------
(0 rows)
```

## Checklist
- [x] My PR name is descriptive, is in imperative tense and starts with one of the following: `[Feature]`,`[Improvement]` or `[Fix]`,
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the RCD team on GitHub, or specific people who are associated with this ticket

### CODE REVIEW

**Review by ChinemeremChigbo — COMMENTED** [review]
I think you should add an Actions column to the requests page to allow for faster request deletion and to match the functionality of applicant deletion.
image

**Comment by sherryhli** [comment]
> I think you should add an Actions column to the requests page to allow for faster request deletion and to match the functionality of applicant deletion. image

@ChinemeremChigbo There isn't an existing actions menu on the requests page so it'll take some more work. I suggest we merge this first and then add it if we have bandwidth in the future, or if RCD asks for it specifically.

File: lib/applications/resolvers.ts [hunk 1]
```diff
@@ -1857,3 +1859,82 @@ export const updateApplicationPhysicianAssessment: Resolver<
...
+
+    const application = await prisma.application.findUnique({
+      where: {
+        id,
+      },
+    });
+
+    if (!application) {
+      return { ok: false, error: `Application with ID ${id} not found` };
+    }
```
Comment (sherryhli) [inline lib/applications/resolvers.ts:1879]: Nit: can use the `rejectOnNotFound` option to throw an error if applicant is not found, this way we can include this fetch inside the try/catch below. Just a style thing, feel free to ignore

```suggestion
    const application = await prisma.application.findUnique({
      where: {
        id,
      },
      rejectOnNotFound: true,
    });

    if (!application) {
      return { ok: false, error: `Application with ID ${id} not found` };
    }
```

**Review by sherryhli — APPROVED** [review]
LGTM

File: lib/applications/resolvers.ts (hunk 1 above)
Comment (leogjhuang) [reply lib/applications/resolvers.ts:1879]: `rejectOnNotFound` appears to be [deprecated](https://www.prisma.io/docs/guides/upgrade-guides/upgrading-versions/upgrading-to-prisma-5/rejectonnotfound-changes) in future Prisma versions, so I might keep this as is for maintainability.

**Comment by ChinemeremChigbo** [comment]
> > I think you should add an Actions column to the requests page to allow for faster request deletion and to match the functionality of applicant deletion. image
> 
> @ChinemeremChigbo There isn't an existing actions menu on the requests page so it'll take some more work. I suggest we merge this first and then add it if we have bandwidth in the future, or if RCD asks for it specifically.

Sound good!

**Comment by leogjhuang** [comment]
Closing this as branch was mistakenly made off `main` instead of `staging`.

**Pull request closed without merging**